### ISSUE TITLE
High CPU usage with encryption

### ISSUE THREAD

**Issue by zfs2019** [openzfsonwindows/ZFSin#110]
With [0.6](https://github.com/openzfsonwindows/ZFSin/releases/tag/0.16), I was be able to mount an encrypted dataset. And I noticed that when I read/write large files, CPU usage goes up to 100%.
Is there no hardware acceleration for crypto?
My CPU supports AES-NI.
```zfs get all pool1```:
```
NAME   PROPERTY              VALUE                  SOURCE
pool1  type                  filesystem             -
pool1  creation              Fri Mar 01 09:47 2019  -
pool1  used                  9.11G                  -
pool1  available             21.9G                  -
pool1  referenced            9.11G                  -
pool1  compressratio         1.00x                  -
pool1  mounted               yes                    -
pool1  quota                 none                   default
pool1  reservation           none                   default
pool1  recordsize            128K                   default
pool1  mountpoint            /pool1                 default
pool1  sharenfs              off                    default
pool1  checksum              on                     default
pool1  compression           on                     local
pool1  atime                 on                     temporary
pool1  devices               on                     default
pool1  exec                  on                     default
pool1  setuid                on                     default
pool1  readonly              off                    default
pool1  zoned                 off                    default
pool1  snapdir               hidden                 default
pool1  aclmode               passthrough            default
pool1  aclinherit            restricted             default
pool1  canmount              on                     default
pool1  xattr                 sa                     local
pool1  copies                1                      default
pool1  version               5                      -
pool1  utf8only              on                     -
pool1  normalization         formD                  -
pool1  casesensitivity       sensitive              -
pool1  vscan                 off                    default
pool1  nbmand                off                    default
pool1  sharesmb              off                    default
pool1  refquota              none                   default
pool1  refreservation        none                   default
pool1  primarycache          all                    default
pool1  secondarycache        all                    default
pool1  usedbysnapshots       0                      -
pool1  usedbydataset         9.11G                  -
pool1  usedbychildren        1.21M                  -
pool1  usedbyrefreservation  0                      -
pool1  logbias               latency                default
pool1  dedup                 off                    default
pool1  mlslabel              none                   default
pool1  sync                  standard               default
pool1  dnodesize             legacy                 default
pool1  refcompressratio      1.00x                  -
pool1  written               9.11G                  -
pool1  logicalused           9.13G                  -
pool1  logicalreferenced     9.13G                  -
pool1  filesystem_limit      none                   default
pool1  snapshot_limit        none                   default
pool1  filesystem_count      none                   default
pool1  snapshot_count        none                   default
pool1  snapdev               hidden                 default
pool1  redundant_metadata    all                    default
pool1  overlay               off                    default
pool1  encryption            aes-128-gcm            -
pool1  keylocation           prompt                 local
pool1  keyformat             passphrase             -
pool1  pbkdf2iters           342K                   -
pool1  encryptionroot        pool1                  -
pool1  keystatus             available              -
pool1  driveletter           -                      default
pool1  special_small_blocks  0                      default
```

**Comment by lundman** [comment]
There is not no, straight C. However, we do have assembler and AES-NI code in the project, but have not got to enabling that on Windows yet - it's on the list.

**Comment by nazar554** [comment]
Is it possible to use Windows kernel crypto API (`CNG`)?

**Comment by lundman** [comment]
Depends, does it do AES-CCM, and AES-GCM, with AAD authenticated data? If so, it's possible. Otherwise, we just need to compile in the asm files, aes-ni and vectorised.

**Comment by nazar554** [comment]
https://docs.microsoft.com/en-us/windows/desktop/seccng/cng-features#kernel-mode-support
https://docs.microsoft.com/en-us/windows/desktop/seccertenroll/cng-cryptographic-algorithm-providers
Not sure about `AAD authenticated data`. It can also do hashing

**Comment by lundman** [comment]
```

CCM (Counter with CBC) | BCRYPT_CHAIN_MODE_CCM | AES | SP800-38C
-- | -- | -- | --
GCM (Galois/Counter Mode) | BCRYPT_CHAIN_MODE_GCM | AES | SP800-38D

```
yes that looks reasonable, AAD is also MAC hash so same same.

**Comment by lundman** [comment]
Since the illumos crypto is pluggable, in theory, making a Windows plugin for it should just slot in an work - that'd be a fun mini project.

**Comment by lundman** [comment]
I left a encrypt/decrypt test case in https://github.com/openzfsonosx/zfs/blob/master/module/icp/illumos-crypto.c#L207

which will help make sure it is correct.

**Comment by nazar554** [comment]
At least encryption stuff seems pluggable, but hashing stuff like `SHA` sometimes gets called directly.

**Comment by lundman** [comment]
the crypt and the hmac is all in icp. But yes, the sha/skein/edonr is baked into ZFS - it shouldnt have been, and there was talk about making those into plugins. But that is separate from crypto - cant use those hashes with crypt.

**Comment by zfs2019** [comment]
Any updates? Sorry for the impatience.
By the way, Microsoft recently [announced](https://devblogs.microsoft.com/commandline/shipping-a-linux-kernel-with-windows/) that Windows 10 will include a native Linux kernel. Now, I'm wondering that if it will enable "ZoL on WSL 2", it will be an alternative for those looking for a way to access ZFS on Windows. Or am I misunderstanding it?

**Comment by lundman** [comment]
I do not know if ZOL will be able to function at first, but presumably eventually it would be able to.

I have looked into the assembly for Windows, and it is a bit of a text file hassle.

We can not use `_asm { }` directive, as it does not support x64.

So it means we have to use the MASM version. Text massaging required at the very least, as MASM can not handle C preprocessor work, so files have to be "simplified" in that sense.

Looking at, for example, 
https://github.com/openzfsonosx/zfs/blob/master/module/icp/asm-x86_64/sha2/sha256_impl.S

All the comments etc, need to start with ";".

Then replace:
```
-ENTRY_NP(SHA256TransformBlocks)
+PUBLIC SHA256TransformBlocks
+;MODEL FLAT, C
+.CODE
+SHA256TransformBlocks PROC
```

Then, we have things like
```
-       push    %rbx
+       push    rbx
```
So, all the `%` used need to be removed, including things like

```
-       lea     (%rsi,%rdx,4),%rdx      # inp+num*16*4
+       lea     [rsi+rdx*4],rdx      ;# inp+num*16*4
```

So, `#` comments don't work, use `;`, then the `(x,y,n)` seems to be `[x+y*n]`.

And that's not even dealing with that the order appears to be different between the assemblers.So that line should probably be
```
+       lea     rdx,[rsi+rdx*4]      ;# inp+num*16*4
```

Minor tweaks like labels can't start with dot;

```
-       jmp     .Lloop
+       jmp     Lloop
-.align 4, 0x90
+align  4
-.Lloop:
+Lloop:
```

and to end it
```
-SET_SIZE(SHA256TransformBlocks)
+SHA256TransformBlocks ENDP

-.align 6, 0x90
+align  8
+.DATA
 K256:
-       .long   0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5
-#endif /* !lint && !__lint */
+       dword   0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5
+
+END
```

So, pretty damned tedious - and all text massaging. Literally anyone could work on this :)

**Comment by zfs2019** [comment]
Can we use the assembly code from [OpenSSL](https://github.com/openssl/openssl)?

**Comment by lundman** [comment]
Yes, certainly the AES, SHA that ZFS uses is already from OpenSSL - we'd have to do it by hand for the  ZFS special code like fletcher and raidz - but that isn't quite as high on desires.

**Comment by zfs2019** [comment]
I would like to enable crypto hardware acceleration, so what can I do?
Do I have to learn MASM? Or just convert the assembly to MASM as you mentioned?

**Comment by lundman** [comment]
Sure, the asm-x86_64/aes/aes_intel.S file is from OpenSSL - so you could fetch the MASM version produced by OpenSSL - which can be used as if - but that will also show the text changes needed to convert the rest of the assembler files that aren't from OpenSSL

**Comment by zfs2019** [comment]
OK, managed to get the [MASM codes from OpenSSL](https://github.com/zfs2019/crypto-win32-masm).

**Comment by lundman** [comment]
Those are the ones - if you compare one to what we already have (gcc/clang version) you can see it is just a "change text around" problem. Probably only needs to be done once. 

However, there are some C files with inline assembler, Windows can not do this, so they'd have to be promoted to full .asm files to be compiled in parallel to the C wrappers.

I spent today porting over ZoL assmebler framework to OSX: https://github.com/openzfsonosx/zfs/pull/715 

which will eventually get to Windows, but does at least show how many assembler files there are;

**Comment by zfs2019** [comment]
Is it possible to have partial hardware acceleration, only using the code from OpenSSL?
So we can reduce the number of work.

**Comment by lundman** [comment]
Yep, openssl has aes, gcm and sha - so that will help you. The extra work would be for fletcher, and raidz.

There is also QAT - which is in ZOL - but we did not port to OsX since there are no drivers - but I guess that assumes you have the hardware required for it.

**Comment by zfs2019** [comment]
Yes, it is only available for some Intel server processors. So lower priority.
By the way, do I have to change the assembler code? Or can it be used as-is?

**Comment by lundman** [comment]
No changes needed - in theory - sometimes there are surprises

**Comment by lundman** [comment]
OK so - having a look again. 

There are two problems here, one is the AT&T syntax vs Intel, ie, Unix vs MS syntax order - which isn't so hard to change. Most toolchains can even specify output type like `gcc -mabi=ms`. Or, converting a single line:

```
# echo "packsswb mm0,[bp+si-0x54]" | llvm-mc-3.2 -x86-asm-syntax=intel
packsswb -84(%bp,%si), %mm0
```

However, the 2nd problem is that the ZFS assembler files are full of pre-processor macros as well, as they are intended to be passed through `cpp`.

So, what are the options. 

* Go full MASM .asm files, and undo the preprocessor logic?
* YASM - can read gas format, but no preprocessor (has --preproc= but cpp no worky)
* Object files? I can build `.o` files on Unix, use `-mabi=ms` and `objconv -fcoff64` to convert to `.obj` file. Then have binary files in github
* clang for VS++ - Now clang is an easy install to VS using the gui installer - if it can do .S files, this could be the answer, if project can be made to skip assembler if clang is not installed.

**Comment by lundman** [comment]
VS2019 Preview and "built in clang" wasn't really what I had hoped for.

But I tried out clang for Windows as a separate install, and can compile `aesni.S` like:
```
# clang -I C:/src/ZFSin/ZFSin/spl/include -c aes_aesni.S -o aes_aesni.obj
```
and drop the resulting `aes_aesni.obj` file into the project, and link against 
`	rijndael_key_setup_enc_intel(NULL, NULL, 0);`

So in theory, a CustomBuildTool perhaps could be made, that uses clang to produce obj file.

Or, we build them once, and include as binary blobs

**Comment by zfs2019** [comment]
Yeah, it seems reasonable, and it makes assembler porting work whole lot easier.
As you mentioned, the inline assembler isn't supported by MSVC for x86_64. So, I used [`__cpuid`](https://docs.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=vs-2017) compiler intrinsic to detect the CPU vendor and AES-NI support. But, I guess it is not needed anymore since clang can do inline assembler.

**Comment by lundman** [comment]
Actually, I do need to find the equivalent for `cpuid_features()` and `cpuid_leaf7_features()` for Windows, so thanks for the start :)

**Comment by lundman** [comment]
I have started on the assembler frame work, clang will now compile both `aes_aesni.S` and `aes_amd64.S` so that is encouraging, however, next up is to complete the SIMD probe, ie, the magic that goes;

https://github.com/openzfsonwindows/ZFSin/blob/zol_raidz/ZFSin/spl/include/sys/simd_x86.h#L311
and down. Fetching values from:

https://github.com/openzfsonwindows/ZFSin/blob/zol_raidz/ZFSin/spl/module/spl/spl-processor.c#L50

**Comment by lundman** [comment]
Ok branch `zol_raidz` is the start of it, its a bit messy since it uses clang to compile the assembler, and C-inline-assembler parts. It does not run well.

`aes_aesni.S` the keygen function will corrupt stack right away, but if I call the C versions of that, so only use assembler for crypto work, it runs for a short while before crashing.

`fletcher4_ssse3` will almost complete a full CrystalMarks run, but also die eventually. 

Presumably they crash because the code trashes a register we should not. I'm not sure. The pure .S files have been changed to use `rcx, rdx, r8, r9` for the first 4 parameters, and that appears correct in the debugger.  It is tempting to turn the pure .S files into C function wrapped ones like fletcher, as that appears to run more easily (no register tweaking).

**Comment by lundman** [comment]
I had some success today with the assembler work, enough to do a little testing.

I don't need to tell anyone to run this only on **test pools**, as it *could* be even more iffy with the assembler code running.

I have tested aesni (ccm+gcm), pclmulqdq (gcm), ssse3 (raidz,fletcher4) and those below are assumed to work; scalar, sse2, x86_64m superscalar4, superscalar. As they are inline, and not stand-alone assembler.

I have **NOT** tested: amd64 (aes), avx512* (fletcher4) as I lack CPU capability.

You can set registry to change the implementation to use, see:
https://openzfsonosx.org/wiki/Windows_Registry#Vectorized_Raidz

Alas, it is only read on boot/install time. It is not dynamic yet. There is no way
to see probed capabilities, except in debug output (DbgView.exe / WinDbg / VisualStudio kernel attach)

Nightly build:
http://www.lundman.net/OpenZFSOnWindows-debug-20190627.exe

Quick benchmarking is somewhat interesting:

CrystalMark 100MiB aes-256-gcm generic
![ZFSin_100MiB_gcm_generic](https://user-images.githubusercontent.com/637168/60246854-86a95100-98fa-11e9-81e9-927243cfa749.png)
CrystalMark 100MiB aes-256-gcm aesni
![ZFSin_100MiB_gcm_aesni](https://user-images.githubusercontent.com/637168/60246867-8c069b80-98fa-11e9-8788-ec8b90511628.png)

It is awkward to benchmark on a VM to start, but there appears to be some difference using AESNI. 

But it is also interesting to note we appear to hit a brick wall after a certain time/amount - this is not related to the assembler work, perhaps to how the ARC behaves once full. But that is a whole other debug session.

**Comment by lundman** [comment]
Note to self: https://github.com/openzfsonosx/zfs/issues/718
Test wide vdevs here as well.

**Comment by zfs2019** [comment]
I have tested the [build](http://www.lundman.net/OpenZFSOnWindows-debug-20190627.exe), but I do still get 100% CPU load. Is hardware acceleration enabled?

**Comment by zfs2019** [comment]
I made the Release build (Debug build is slower) and did some benchmarks.

Here are some interesting results:
![1561846805](https://user-images.githubusercontent.com/48113574/60390044-8c748180-9b08-11e9-90a2-c2ce1b2bd730.png)
![1561847002](https://user-images.githubusercontent.com/48113574/60390045-8c748180-9b08-11e9-9fe7-18eca3fdc4ef.png)

Notice the higher CPU load of `gcm`
I think there is something wrong with `gcm`

Notes:
* I disabled cachemgr in this build
* I set:
  * `zfs_arc_max` to `268435456`
  * `icp_aes_impl` to `aesni`
  * `icp_gcm_impl` to `fastest`

**Comment by lundman** [comment]
Hmm that is interesting - well done on building it, I've not got to updating the build instructions :)

**Comment by zfs2019** [comment]
Yeah, it was a bit hassle :)
It attracts me into `ccm` since it is faster.

**Comment by lundman** [comment]
Should look at that weird thing that happens after a bit and writes go to hell.

**Comment by zfs2019** [comment]
OK, I found the cause of `gcm` slowdown.
The `pclmulqdq` instruction was not being used.

So, by changing:
https://github.com/openzfsonwindows/ZFSin/blob/ef4f7108863c3d746af5f63fa2fe96e93cd1ab16/ZFSin/zfs/module/icp/algs/modes/gcm_pclmulqdq.c#L54-L62
To:
```c
static boolean_t
gcm_pclmulqdq_will_work(void)
{
	return (zfs_pclmulqdq_available());
}
```
I was able to get acceleration:
![1561886137](https://user-images.githubusercontent.com/48113574/60394717-c0c95b80-9b63-11e9-956e-8a4f1d5bae94.png)
![1561886294](https://user-images.githubusercontent.com/48113574/60394718-c0c95b80-9b63-11e9-877a-d52a42d93738.png)

**Comment by lundman** [comment]
Hah yep - that is right, that was left in there after doing the wrong thing. That file is only compiled by VC++

**Comment by zfs2019** [comment]
Please update that file :)

**Comment by zfs2019** [comment]
So the fletcher4 was, too, not getting `sse2` and `ssse3` acceleration because of `__clang__` in https://github.com/openzfsonwindows/ZFSin/blob/ef4f7108863c3d746af5f63fa2fe96e93cd1ab16/ZFSin/zfs/module/zcommon/zfs_fletcher_sse.c?

**Comment by lundman** [comment]
Whomever checked that code in clearly did a shoddy job eh! I'll have it fixed up tomorrow.

**Comment by zfs2019** [comment]
Oh, I have a question. Are AVX and AVX2 acceleration of fletcher4 available?

**Comment by lundman** [comment]
I fixed up `gcm_pclmulqdq` to be linked, but I don't think fletcher_sse2 was affected as it is built by clang. 
As for AVX and AVX2 - they are compiled in, but I have no way to confirm I got the word order correct for looking up leaf7 bits:
https://github.com/openzfsonwindows/ZFSin/blob/zol_raidz/ZFSin/spl/module/spl/spl-processor.c#L70
https://github.com/openzfsonwindows/ZFSin/blob/zol_raidz/ZFSin/spl/include/sys/simd_x86.h#L261

**Comment by zfs2019** [comment]
By the way, if you try to build in Release, you will get errors like:
![ZFSin - Microsoft Visual Studio](https://user-images.githubusercontent.com/48113574/60546118-a48e1000-9d57-11e9-9646-b36bc1eba682.png)
which are probably caused by a different behavior between Debug and Release builds.
I workaround by changing `gcm_generic.c` like:
```diff
@@ -24,6 +24,9 @@

 #include <modes/gcm_impl.h>

+#define htonll(x)       _byteswap_uint64(x)
+#define ntohll(x)       _byteswap_uint64(x)
+
 struct aes_block {
 	uint64_t a;
 	uint64_t b;
```
It works, but what the correct solution would be?

**Comment by lundman** [comment]
Thanks for the heads up, I have cleaned that up I think.

**Comment by zfs2019** [comment]
Thanks, I confirmed it works.

**Comment by zfs2019** [comment]
Could you please re-base with the master and release the signed installer? So I don't have to enable Test Mode.  
BTW, I'm using it almost every day for the past several months. There were BSoDs but **no data loss** so far and it works fine, at least for me.

**Comment by lundman** [comment]
7d8c1ac

http://www.lundman.net/OpenZFSOnWindows-release-20191123-ASM.exe

**Comment by zfs2019** [comment]
Thank you @lundman, for your effort. I appreciate it.
It's working fine!

![Windows｜1574410280](https://user-images.githubusercontent.com/48113574/69409598-9b283300-0d4c-11ea-9485-604d9e697741.png)
